### PR TITLE
feat: Adding the Terraform module to deploy SD-Core Bundles

### PR DESCRIPTION
**Pull request by gatici**
Adding the Terraform module to deploy
-  SD-Core Control Plane
-  SD-Core User Plane
-  SD-Core

Note:
This repository is aimed to keep `sdcore-control-plane`, `sdcore-control-plane` and `sdcore ` TF modules. 
This PR is scoped to add only the `sdcore-control-plane` TF module.

### CODE REVIEW

File: main.tf
[the remark this answers is not in the document]
Comment (gatici) [reply main.tf:1]: Yes, that's possible. There is one drawback of this. If we would like to publish them in [Public Terraform Registry](https://registry.terraform.io/), they should be in the separate repositories.

File: LICENSE [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+      To apply the Apache License to your work, attach the following
+      boilerplate notice, with the fields enclosed by brackets "[]"
+      replaced with your own identifying information. (Don't include
+      the brackets!)  The text should be enclosed in the appropriate
+      comment syntax for the file format. We also recommend that a
+      file or class name and description of purpose be included on the
+      same "printed page" as the copyright notice for easier
+      identification within third-party archives.
+
+   Copyright 2023 Canonical Ltd.
```
[the remark this answers is not in the document]
Comment (gatici) [reply LICENSE]: Done

File: main.tf
[the remark this answers is not in the document]
Comment (gatici) [reply main.tf:1]: Done

[the remark this answers is not in the document]
Comment (gatici) [reply main.tf:1]: Repository structure is updated to keep `sdcore`, `sdcore-control-plane` and `sdcore-user-plane` modules.

File: modules/sdcore-control-plane-k8s/main.tf [hunk 2]
```diff
@@ -0,0 +1,176 @@
+resource "juju_model" "sdcore" {
+  name = var.model_name
+}
+
+module "sdcore-amf-k8s" {
+  source                 = "git::https://github.com/canonical/sdcore-amf-k8s-operator//terraform"
```
Comment (markbeierl) [inline modules/sdcore-control-plane-k8s/main.tf:6]: Are the two // needed?  `sdcore-amf-k8s-operator//terraform` vs `sdcore-amf-k8s-operator/terraform`.  All the sources have that

File: modules/sdcore-control-plane-k8s/main.tf [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+  source                         = "git::https://github.com/canonical/grafana-agent-k8s-operator//terraform"
+  model_name                     = juju_model.sdcore.name
+  channel                        = var.grafana-channel
+  metrics_remote_write_offer_url = var.metrics_remote_write_offer_url
+  logging_offer_url              = var.logging_offer_url
+  grafana-config                 = var.grafana-config
+}
+
+module "self-signed-certificates" {
+  source      = "git::https://github.com/canonical/self-signed-certificates-operator.git//terraform"
```
Comment (markbeierl) [inline modules/sdcore-control-plane-k8s/main.tf:107]: Source?

File: modules/sdcore-control-plane-k8s/main.tf [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+  model_name  = juju_model.sdcore.name
+  channel     = var.cert-channel
+  cert-config = var.cert-config
+}
+
+module "traefik-k8s" {
+  source         = "git::https://github.com/canonical/traefik-k8s-operator//terraform"
```
Comment (markbeierl) [inline modules/sdcore-control-plane-k8s/main.tf:114]: Source?

File: modules/sdcore-control-plane-k8s/main.tf (hunk 2 above)
Comment (gatici) [reply modules/sdcore-control-plane-k8s/main.tf:6]: We need  use sdcore-amf-k8s-operator//terraform as the modules are kept under a subdirectory. Please check:  https://developer.hashicorp.com/terraform/language/modules/sources#modules-in-package-sub-directories

File: modules/sdcore-control-plane-k8s/main.tf [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+module "mongodb-k8s" {
+  source       = "git::https://github.com/canonical/mongodb-operator.git//terraform"
+  model_name   = juju_model.sdcore.name
+  channel      = var.mongo-channel
+  mongo-config = var.mongo-config
+}
+
+module "grafana-agent-k8s" {
+  source                         = "git::https://github.com/canonical/grafana-agent-k8s-operator//terraform"
```
[the remark this answers is not in the document]
Comment (gatici) [reply modules/sdcore-control-plane-k8s/main.tf:98]: It's same for all the charms. The source is `terraform` subdirectory in the main branch of charm.

File: modules/sdcore-control-plane-k8s/main.tf (hunk 3 above)
Comment (gatici) [reply modules/sdcore-control-plane-k8s/main.tf:107]: It's same for all the charms. The source is `terraform` subdirectory in the main branch of charm.

File: modules/sdcore-control-plane-k8s/main.tf (hunk 4 above)
Comment (gatici) [reply modules/sdcore-control-plane-k8s/main.tf:114]: It's same for all the charms. The source is terraform subdirectory in the main branch of charm.

File: modules/sdcore-control-plane-k8s/main.tf (hunk 5 above)
[the remark this answers is not in the document]
Comment (markbeierl) [reply modules/sdcore-control-plane-k8s/main.tf:98]: I'm looking for the PR for this

File: modules/sdcore-control-plane-k8s/main.tf (hunk 2 above)
Comment (markbeierl) [reply modules/sdcore-control-plane-k8s/main.tf:6]: Understood: 
> A special double-slash syntax is interpreted by Terraform

File: .github/workflows/issues.yaml [hunk 6]
```diff
@@ -0,0 +1,11 @@
+name: Sync Issues to JIRA
+
+on:
+  issues:
+    types: [opened, reopened, closed]
+
+jobs:
+  update:
+    name: Update Issue
+    uses: canonical/sdcore-github-workflows/.github/workflows/issues.yaml@main
```
Comment (Gmerold) [inline .github/workflows/issues.yaml:10]: This action requires `JIRA_URL` secret to be added in the project settings. Unfortunately I'd exactly remember the value. Perhaps Yazan does.

File: .github/workflows/terraform.yaml [hunk 7]
```diff
@@ -0,0 +1,30 @@
+name: Check the Terraform Modules
```
Comment (Gmerold) [inline .github/workflows/terraform.yaml:1]: I think this workflow has already been moved to the `sdcore-github-workflows` project, right?

**Review by Gmerold — COMMENTED** [review]
Are we going to store all high level modules in this repo? I mean SD-Core all-in-one, CP and UP?

File: .github/workflows/terraform.yaml (hunk 7 above)
Comment (gatici) [reply .github/workflows/terraform.yaml:1]: yes, it's been removed. This repo is not updated recently. I will update it.

File: .github/workflows/issues.yaml (hunk 6 above)
Comment (gatici) [reply .github/workflows/issues.yaml:10]: Yes, we will store all high levels SD-Core modules here.

File: .github/workflows/terraform.yaml (hunk 7 above)
Comment (gatici) [reply .github/workflows/terraform.yaml:1]: Done

**Review by Gmerold — COMMENTED** [review]
I think all the base modules are merged now (at least SD-Core ones), so we can probably update this PR.

File: .github/workflows/issues.yaml (hunk 6 above)
Comment (gatici) [reply .github/workflows/issues.yaml:13]: Added.

File: modules/sdcore-control-plane-k8s/README.md [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+```yaml
+# Mandatory Config Options
+model_name = "put your model-name here"
+
+# Optional Configurations
+
+# channel                        = "put the channel for the SD-Core charms here"
+# mongo_channel                  = "put the MongoDB charm channel here"
+# grafana_channel                = "put the Grafana charm channel here"
```
Comment (Gmerold) [inline modules/sdcore-control-plane-k8s/README.md:44]: It's Grafana agent actually. It's like a worker that collects the data for Grafana.

File: modules/sdcore-control-plane-k8s/README.md [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+# self_signed_certificates_channel                   = "put the Self Signed Certificates charm channel here"
+# traefik_channel                = "put the Self Signed Certificates charm channel here"
```
Comment (Gmerold) [inline modules/sdcore-control-plane-k8s/README.md:46]:
```suggestion
# traefik_channel                = "put the Traefik charm channel here"
```

File: modules/sdcore-control-plane-k8s/README.md [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+- **variables.tf** - Allows customization of the deployment. Except for exposing the deployment options (Juju model name, channel) also allows overwriting charm's default configuration.
+- **terraform.tf** - Defines the Terraform provider.
+
+## Deploying sdcore-control-plane-k8s module separately
+
+### Pre-requisites
+
+The following tools needs to be installed and should be running in the environment. Please [set up your environment][set-up-environment] before deployment.
+
+- A Kubernetes cluster with the `Multus` and `Metallb` addon enabled.
```
Comment (Gmerold) [inline modules/sdcore-control-plane-k8s/README.md:21]: Should we also mention the required number of addresses load balancer should have?

File: modules/sdcore-control-plane-k8s/main.tf [hunk 11]
```diff
@@ -0,0 +1,485 @@
+resource "juju_model" "sdcore" {
+  name = var.model_name
+}
+
+module "amf" {
+  source     = "git::https://github.com/canonical/sdcore-amf-k8s-operator//terraform"
+  model_name = juju_model.sdcore.name
+  channel    = var.channel
+  amf-config = var.amf_config
```
Comment (Gmerold) [inline modules/sdcore-control-plane-k8s/main.tf:9]:
```suggestion
  config = var.amf_config
```

File: modules/sdcore-control-plane-k8s/main.tf [hunk 12]
```diff
@@ -0,0 +1,485 @@
...
+  source     = "git::https://github.com/canonical/sdcore-nrf-k8s-operator//terraform"
+  model_name = juju_model.sdcore.name
+  channel    = var.channel
+}
+
+module "nssf" {
+  source      = "git::https://github.com/canonical/sdcore-nssf-k8s-operator//terraform"
+  model_name  = juju_model.sdcore.name
+  channel     = var.channel
+  nssf-config = var.nssf_config
```
Comment (Gmerold) [inline modules/sdcore-control-plane-k8s/main.tf:34]:
```suggestion
  config = var.nssf_config
```

File: modules/sdcore-control-plane-k8s/main.tf [hunk 13]
```diff
@@ -0,0 +1,485 @@
...
+  model_name   = juju_model.sdcore.name
+  channel      = var.mongo_channel
+  mongo-config = var.mongo_config
+}
+
+module "grafana-agent" {
+  source         = "git::https://github.com/canonical/grafana-agent-k8s-operator//terraform"
+  model_name     = juju_model.sdcore.name
+  channel        = var.grafana_channel
+  grafana-config = var.grafana_config
```
Comment (Gmerold) [inline modules/sdcore-control-plane-k8s/main.tf:78]:
```suggestion
  config = var.grafana_agent_config
```

File: modules/sdcore-control-plane-k8s/main.tf [hunk 14]
```diff
@@ -0,0 +1,485 @@
...
+}
+
+module "self-signed-certificates" {
+  source      = "git::https://github.com/canonical/self-signed-certificates-operator.git//terraform"
+  model_name  = juju_model.sdcore.name
+  channel     = var.self_signed_certificates_channel
+  cert-config = var.self_signed_certificates_config
```
Comment (Gmerold) [inline modules/sdcore-control-plane-k8s/main.tf:85]:
```suggestion
  config = var.self_signed_certificates_config
```

File: modules/sdcore-control-plane-k8s/main.tf [hunk 15]
```diff
@@ -0,0 +1,485 @@
...
+}
+
+module "traefik" {
+  source         = "git::https://github.com/canonical/traefik-k8s-operator//terraform"
+  model_name     = juju_model.sdcore.name
+  channel        = var.traefik_channel
+  traefik-config = var.traefik_config
```
Comment (Gmerold) [inline modules/sdcore-control-plane-k8s/main.tf:92]:
```suggestion
  config = var.traefik_config
```

**Review by Gmerold — CHANGES_REQUESTED** [review]

File: modules/sdcore-control-plane-k8s/main.tf [hunk 16]
```diff
@@ -0,0 +1,191 @@
+resource "juju_model" "sdcore" {
```
[the remark this answers is not in the document]
Comment (gatici) [reply modules/sdcore-control-plane-k8s/main.tf:1]: Done

File: modules/sdcore-control-plane-k8s/variables.tf [hunk 17]
```diff
@@ -0,0 +1,79 @@
+variable "model_name" {
```
[the remark this answers is not in the document]
Comment (gatici) [reply modules/sdcore-control-plane-k8s/variables.tf:4]: Done

File: README.md [hunk 18]
```diff
@@ -0,0 +1,110 @@
+# SD-Core Terraform Module
+
+This folder contains the [Terraform][Terraform] module for [sdcore-k8s][sdcore-k8s] consisting of [sdcore-control-plane-k8s][sdcore-control-plane-k8s] and [sdcore-user-plane-k8s][sdcore-user-plane-k8s] modules.
+
+The modules use the [Terraform Juju provider][Terraform Juju provider] to model the charm deployment onto any Kubernetes environment managed by [Juju][Juju].
+
+## Module structure
+
+- **main.tf** - Defines the Juju application to be deployed.
+- **variables.tf** - Allows customization of the deployment. Except for exposing the deployment options (Juju model name, channel) also allows overwriting charm's default configuration.
```
Comment (markbeierl) [inline README.md:10]: I'm not quite sure what this statement means

File: CONTRIBUTING.md [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+
+### Prerequisites
+
+Make sure the following software and tools are installed in the development environment.
+
+- `microk8s`
+- `juju`
+- `terraform`
+
+### Prepare Development Environment
```
[the remark this answers is not in the document]
Comment (gatici) [reply CONTRIBUTING.md]: Ok, this part is removed.

File: README.md [hunk 20]
```diff
@@ -0,0 +1,13 @@
+# SD-Core Terraform Module
```
[the remark this answers is not in the document]
Comment (gatici) [reply README.md:1]: Added.

File: modules/sdcore-control-plane-k8s/README.md [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+- `microk8s`
+- `juju 3.x`
+- `terrafom`
+
+### Deploy the sdcore-control-plane-k8s bundle using Terraform
+
+Make sure that `storage`, `multus` and `metallb` plugins are enabled for Microk8s:
+
+```console
+sudo microk8s enable hostpath-storage multus
```
[the remark this answers is not in the document]
Comment (gatici) [reply modules/sdcore-control-plane-k8s/README.md]: Added.

File: modules/sdcore-control-plane-k8s/README.md [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+
+## Getting Started
+
+### Prerequisites
+
+The following software and tools needs to be installed and should be running in the local environment. Please [set up your environment](https://discourse.charmhub.io/t/set-up-your-development-environment-with-microk8s-for-juju-terraform-provider/13109) before deployment.
+
+- `microk8s`
+- `juju 3.x`
+- `terrafom`
```
[the remark this answers is not in the document]
Comment (gatici) [reply modules/sdcore-control-plane-k8s/README.md]: Done

File: modules/sdcore-control-plane-k8s/README.md [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+
+### Deploy the sdcore-control-plane-k8s bundle using Terraform
+
+Make sure that `storage`, `multus` and `metallb` plugins are enabled for Microk8s:
```
[the remark this answers is not in the document]
Comment (gatici) [reply modules/sdcore-control-plane-k8s/README.md]: Done

File: modules/sdcore-control-plane-k8s/README.md [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+
+```console
+sudo microk8s enable hostpath-storage multus
+sudo microk8s enable metallb:10.0.0.2-10.0.0.4
```
[the remark this answers is not in the document]
Comment (gatici) [reply modules/sdcore-control-plane-k8s/README.md]: Done

File: modules/sdcore-control-plane-k8s/README.md [hunk 25]
```diff
@@ -0,0 +1,63 @@
...
+```
+
+Initialise the provider:
+
+```console
+terraform init
+```
+
+Fill the mandatory config options in the `terraform.tfvars` file.  The provided model-name is not expected to pre-exist and will be created by Juju Terraform Provider.
```
[the remark this answers is not in the document]
Comment (gatici) [reply modules/sdcore-control-plane-k8s/README.md]: Done

File: CONTRIBUTING.md [hunk 26]
```diff
@@ -0,0 +1,80 @@
+# Contributing
+
+## Prerequisites
+
+To make contributions to this repository, the following software and tools needs to be installed in your development environment.
```
[the remark this answers is not in the document]
Comment (gatici) [reply CONTRIBUTING.md:5]: Done

File: CONTRIBUTING.md [hunk 27]
```diff
@@ -0,0 +1,80 @@
...
+
+```console
+juju bootstrap microk8s
+```
+
+Add a Juju model:
+
+```console
+juju add-model <model-name>
+````
```
[the remark this answers is not in the document]
Comment (gatici) [reply CONTRIBUTING.md:50]: Removed.

File: CONTRIBUTING.md [hunk 28]
```diff
@@ -0,0 +1,80 @@
...
+
+A Terraform working directory needs to be initialized at the beginning.
+
+Initialise the provider:
+
+```console
+terraform init
+```
+
+Formats the *.tf files to a canonical format and style:
```
[the remark this answers is not in the document]
Comment (gatici) [reply CONTRIBUTING.md:64]: Done

File: CONTRIBUTING.md [hunk 29]
```diff
@@ -0,0 +1,80 @@
...
+
+```console
+terraform fmt
+```
+
+Check the syntactical validation:
```
[the remark this answers is not in the document]
Comment (gatici) [reply CONTRIBUTING.md:70]: I prefered `check the syntax`.

File: README.md [hunk 30]
```diff
@@ -0,0 +1,33 @@
...
+
+This folder contains the [Terraform][Terraform] module for the [sdcore-k8s][sdcore-k8s].
```
[the remark this answers is not in the document]
Comment (gatici) [reply README.md]: We gave reference to the bundles.

File: modules/sdcore-control-plane-k8s/README.md [hunk 31]
```diff
@@ -0,0 +1,103 @@
+# SD-Core Control Plane Terraform Module
+
+This folder contains the [Terraform][Terraform] module for the [sdcore-control-plane-k8s][sdcore-control-plane-k8s].
```
[the remark this answers is not in the document]
Comment (gatici) [reply modules/sdcore-control-plane-k8s/README.md]: Done

File: modules/sdcore-control-plane-k8s/README.md [hunk 32]
```diff
@@ -0,0 +1,103 @@
...
+
+The module uses the [Terraform Juju provider][Terraform Juju provider] to model the charm deployment onto any Kubernetes environment managed by [Juju][Juju].
+
+The module can be used to deploy the `sdcore-control-plane-k8s` separately as well as a part of `sdcore-k8s` module, depending on the deployment architecture.
```
[the remark this answers is not in the document]
Comment (gatici) [reply modules/sdcore-control-plane-k8s/README.md:7]: I replaced it with `as well as a part of the higher level modules`

File: modules/sdcore-control-plane-k8s/README.md (hunk 8 above)
Comment (gatici) [reply modules/sdcore-control-plane-k8s/README.md]: It's being replaced as `grafana-agent`

File: modules/sdcore-control-plane-k8s/README.md (hunk 9 above)
Comment (gatici) [reply modules/sdcore-control-plane-k8s/README.md]: Those disabled config options are removed.

File: modules/sdcore-control-plane-k8s/README.md (hunk 10 above)
Comment (gatici) [reply modules/sdcore-control-plane-k8s/README.md:21]: Required number of addresses for load balancer is added to the requirements

File: modules/sdcore-control-plane-k8s/main.tf (hunk 11 above)
Comment (gatici) [reply modules/sdcore-control-plane-k8s/main.tf:12]: Corrected.

File: modules/sdcore-control-plane-k8s/main.tf (hunk 12 above)
Comment (gatici) [reply modules/sdcore-control-plane-k8s/main.tf:37]: Corrected.

File: modules/sdcore-control-plane-k8s/main.tf (hunk 13 above)
Comment (gatici) [reply modules/sdcore-control-plane-k8s/main.tf]: Corrected.

File: modules/sdcore-control-plane-k8s/main.tf (hunk 14 above)
Comment (gatici) [reply modules/sdcore-control-plane-k8s/main.tf:88]: Corrected.

File: modules/sdcore-control-plane-k8s/main.tf (hunk 15 above)
Comment (gatici) [reply modules/sdcore-control-plane-k8s/main.tf:95]: Corrected.

File: README.md (hunk 18 above)
Comment (gatici) [reply README.md:10]: This statement is replaced with `Allows customization of the deployment including Juju model name, charm's channel and configuration.`

File: modules/sdcore-control-plane-k8s/README.md [hunk 33]
```diff
@@ -0,0 +1,63 @@
...
+terraform plan -var-file="terraform.tfvars" 
+```
+
+Deploy the resources:
+
+```console
+terraform apply -auto-approve 
+```
+
+### Check the Output
```
[the remark this answers is not in the document]
Comment (gatici) [reply modules/sdcore-control-plane-k8s/README.md:91]: I do not have strong opinion here. We can.

File: modules/sdcore-control-plane-k8s/variables.tf [hunk 34]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+variable "channel" {
+  description = "The channel to use when deploying `SD-Core` charms."
+  type        = string
+  default     = "1.3/edge"
+}
+
+variable "mongo_channel" {
+  description = "The channel to use when deploying `mongodb-k8s` charm "
```
Comment (Gmerold) [inline modules/sdcore-control-plane-k8s/variables.tf:17]:
```suggestion
  description = "The channel to use when deploying `mongodb-k8s` charm."
```

File: modules/sdcore-control-plane-k8s/variables.tf [hunk 35]
```diff
@@ -0,0 +1,76 @@
...
+  type        = string
+  default     = "6/beta"
+}
+
+variable "grafana_agent_channel" {
+  description = "The channel to use when deploying `grafana-agent-k8s` charm "
```
Comment (Gmerold) [inline modules/sdcore-control-plane-k8s/variables.tf:23]:
```suggestion
  description = "The channel to use when deploying `grafana-agent-k8s` charm."
```

File: modules/sdcore-control-plane-k8s/variables.tf [hunk 36]
```diff
@@ -0,0 +1,76 @@
...
+  type        = string
+  default     = "latest/stable"
+}
+
+variable "self_signed_certificates_channel" {
+  description = "The channel to use when deploying `self-signed-certificates-operator` charm "
```
Comment (Gmerold) [inline modules/sdcore-control-plane-k8s/variables.tf:29]:
```suggestion
  description = "The channel to use when deploying `self-signed-certificates-operator` charm."
```

File: modules/sdcore-user-plane-k8s/README.md [hunk 37]
```diff
@@ -0,0 +1,124 @@
+# SD-Core Control Plane Terraform Module
```
Comment (Gmerold) [inline modules/sdcore-user-plane-k8s/README.md:1]:
```suggestion
# SD-Core User Plane Terraform Module
```

File: modules/sdcore-user-plane-k8s/README.md [hunk 38]
```diff
@@ -0,0 +1,124 @@
...
+```shell
+sudo microk8s addons repo add community https://github.com/canonical/microk8s-community-addons --reference feat/strict-fix-multus
+```
+
+Enable the `hostpath-storage`, `multus` and `metallb` MicroK8s addons.
+
+```shell
+sudo microk8s enable hostpath-storage
+sudo microk8s enable multus
+sudo microk8s enable metallb:10.0.0.2-10.0.0.2
```
Comment (Gmerold) [inline modules/sdcore-user-plane-k8s/README.md:49]:
```suggestion
sudo microk8s enable metallb:10.0.0.2/32
```

File: main.tf [hunk 39]
```diff
@@ -0,0 +1,509 @@
+# Copyright 2024 Canonical Ltd.
```
Comment (Gmerold) [inline main.tf:1]: It feels a bit odd to have this module in the main folder while other SD-Core modules are stored under `modules`. Especially in the situation where this module doesn't reuse `sdcore-control-plane-k8s` and `sdcore-user-plane-k8s`.
How about we move it together with the rest of the modules?

File: modules/sdcore-control-plane-k8s/README.md (hunk 33 above)
[the remark this answers is not in the document]
Comment (Gmerold) [reply modules/sdcore-control-plane-k8s/README.md:91]: OK, let's skip it. We cover this part in the tutorial anyway.

**Review by Gmerold — COMMENTED** [review]
One general comment - I think each module should have `output.tf` to expose the names of the applications that can have external integrations and the endpoints for those integrations. For the CP module that would be AMF (fivef-n2) and NMS (fiveg_gnb_identity).
For the UP module UPF with its fiveg-n3 and fiveg-n4 interfaces.

File: modules/sdcore-control-plane-k8s/variables.tf (hunk 34 above)
Comment (gatici) [reply modules/sdcore-control-plane-k8s/variables.tf]: Done

File: modules/sdcore-control-plane-k8s/variables.tf (hunk 35 above)
Comment (gatici) [reply modules/sdcore-control-plane-k8s/variables.tf]: Done

File: modules/sdcore-control-plane-k8s/variables.tf (hunk 36 above)
Comment (gatici) [reply modules/sdcore-control-plane-k8s/variables.tf]: Done

File: modules/sdcore-user-plane-k8s/README.md (hunk 38 above)
Comment (gatici) [reply modules/sdcore-user-plane-k8s/README.md]: Done

**Comment by gatici** [comment]
> One general comment - I think each module should have `output.tf` to expose the names of the applications that can have external integrations and the endpoints for those integrations. For the CP module that would be AMF (fivef-n2), NMS (fiveg_gnb_identity) and grafana_agent (send-remote-write and logging-consumer). For the UP module that should be UPF (fiveg-n3 and fiveg-n4) and grafana_agent (send-remote-write and logging-consumer). For all-in-one combination of both.

The output.tf files are added for all the sdcore modules.

File: modules/sdcore-control-plane-k8s/outputs.tf [hunk 40]
```diff
@@ -0,0 +1,17 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+output "amf_app_name" {
```
Comment (Gmerold) [inline modules/sdcore-control-plane-k8s/outputs.tf:4]: What about integration endpoints?
```
output "fiveg_n2_endpoint" {
  description = "Name of the endpoint used to provide information on connectivity to the N2 plane."
  value       = "fiveg_n2_endpoint"
}
And so on...

File: modules/sdcore-k8s/README.md [hunk 41]
```diff
@@ -0,0 +1,192 @@
+# SD-Core Terraform Module
+
+This project contains 3 [Terraform][Terraform] modules to deploy the following SD-Core bundles: [SD-Core][sdcore-k8s], [SD-Core CP][sdcore-control-plane-k8s] and [SD-Core UP][sdcore-user-plane-k8s].
+
+The modules use the [Terraform Juju provider][Terraform Juju provider] to model the bundle deployment onto any Kubernetes environment managed by [Juju][Juju].
+
+`sdcore-k8s` module deploys a standalone 5G core network. This module contains the 5G control plane functions, the UPF, Webui, Grafana Agent, Traefik, Self Signed Certificates and MongoDB.
+
+`sdcore-control-plane-k8s` module deploys the 5G control plane and the `sdcore-user-plane-k8s` module deploys only the 5G user plane. Hence, the SD-Core 5G core network deployment could be performed following the Control and User Plane Separation (CUPS) principles.
```
Comment (Gmerold) [inline modules/sdcore-k8s/README.md:9]: I guess this should rather go to the README in the root of the repository, right?

File: modules/sdcore-user-plane-k8s/outputs.tf [hunk 42]
```diff
@@ -0,0 +1,12 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+output "upf_app_name" {
```
Comment (Gmerold) [inline modules/sdcore-user-plane-k8s/outputs.tf:4]: Same as above. I think we should also have available endpoints here.

**Review by Gmerold — CHANGES_REQUESTED** [review]

File: modules/sdcore-control-plane-k8s/outputs.tf (hunk 40 above)
Comment (gatici) [reply modules/sdcore-control-plane-k8s/outputs.tf:4]: The required integration points are added.

File: modules/sdcore-k8s/README.md (hunk 41 above)
Comment (gatici) [reply modules/sdcore-k8s/README.md]: This part is moved to the main folder.

File: modules/sdcore-user-plane-k8s/outputs.tf (hunk 42 above)
Comment (gatici) [reply modules/sdcore-user-plane-k8s/outputs.tf:4]: Done.

**Review by Gmerold — APPROVED** [review]
That looks good to me. Well done @gatici!

**Comment by gatici** [comment]
Terraform check action fails because of default TF modules path. This will be fixed after merging: https://github.com/canonical/sdcore-github-workflows/pull/31

**Comment by gatici** [comment]
> That looks good to me. Well done @gatici!

Thanks for your great help and guidance @Gmerold !!

**Pull request merged**